src/mapped_tcp_socket.rs

```diff
@@ -783,33 +792,112 @@ pub fn tcp_punch_hole(socket: net2::TcpBuilder,
+
+                let mut other_streams = Vec::new();
+                // Collect any other successful connections that may have occured simultaneously to
+                // the one we've got.
+                for stream_res_opt in results_rx {
+                    match stream_res_opt {
+                        // Just the timer thread shutting down.
+                        None => (),
+                        // A worker thread had an error. But we don't care because we've already
+                        // got a successful connection.
+                        Some(Err(_)) => (),
+                        // We made another connection.
+                        Some(Ok((stream, stream_addr))) => other_streams.push((stream, stream_addr)),
+                    };
+                }
+
+                if other_streams.len() == 0 {
+                    return WOk(stream, warnings);
+                }
+                else {
+                    let mut errors = Vec::new();
+                    other_streams.push((stream, stream_addr));
+                    let streams: Vec<(TcpStream, SocketAddr, u64)> = other_streams.into_iter().filter_map(|(mut stream, stream_addr)| {
+                        let w = random();
+                        match stream.write_u64::<BigEndian>(w) {
+                            Ok(()) => (),
+                            Err(e) => {
+                                errors.push(TcpPunchHoleBrokenStream {
+                                    peer_addr: stream_addr,
+                                    error: e,
+                                });
+                                return None;
+                            },
+                        };
+                        Some((stream, stream_addr, w))
+                    }).collect();
+                    let stream_opt = streams.into_iter().fold(None, |opt, (mut this_stream, this_stream_addr, w)| {
+                        let this_sum = match this_stream.read_u64::<BigEndian>() {
+                            Ok(r) => r.wrapping_add(w),
+                            Err(e) => {
+                                errors.push(TcpPunchHoleBrokenStream {
+                                    peer_addr: this_stream_addr,
+                                    error: e,
+                                });
+                                return opt;
+                            },
+                        };
+                        match opt {
+                            Some((top_stream, top_sum)) => {
+                                if this_sum > top_sum {
+                                    Some((this_stream, this_sum))
+                                }
+                                else {
+                                    Some((top_stream, top_sum))
+                                }
+                            },
+                            None => Some((this_stream, this_sum))
+                        }
+                    });
```

Comment:
Care to elaborate what you're doing here? Are you dropping all but the last responsive stream? A comment explaining a little of context why this is needed would help a lot.